File: 序号/18.四数之和.kt
fun fourSum(nums: IntArray, target: Int): List<List<Int>> {
    // 4数之和，和两数之和、三数之和一样，暴力需要n^4,利用两数之和的降低一次幂，排序需要logN,总时间N^3+logN
    // 固定2个，右边两个向中间缩进查找
    val n = nums.size
    nums.sort()
    val set = mutableSetOf<String>()
    var a = 0 //4个指针用abcd
    while (a <= n - 4) {
        var b = a + 1
        while (b <= n - 3) {
            var c = b + 1
            var d = n - 1
            val nA = nums[a]
            val nB = nums[b]
            while (c < d) {
                val nC = nums[c]
                val nD = nums[d]
                val t = target - nA - nB - nC - nD
                if (t == 0) {
                    set.add("${nA}_${nB}_${nC}_${nD}")
                    c++
                    d--
                } else if (t > 0) {
                    // nc+nd 小了
                    c++
                } else if (t < 0) {
                    d--
                }
            }
            b++
        }
        a++
    }
    return set.map { str->str.split("_").map { numStr->numStr.toInt() } }
}

File: 序号/18. 四数之和.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        n=len(nums)
        ans=[]
        i=0
        while i<n-3:
            if i and nums[i]==nums[i-1]:
                i+=1
                continue
            if nums[i]+nums[-1]+nums[-2]+nums[-3]<target:
                i+=1
                continue
            if nums[i]+nums[i+1]+nums[i+2]+nums[i+3]>target:
                break
            j=i+1
            while j<n-2:
                if j!=i+1 and nums[j]==nums[j-1]:
                    j+=1
                    continue
                if nums[i]+nums[j]+nums[j+1]+nums[j+2]>target:
                    break
                if nums[i]+nums[j]+nums[-1]+nums[-2]<target:
                    j+=1
                    continue
                l,r=j+1,n-1
                while l<r:
                    s=nums[i]+nums[j]+nums[l]+nums[r]
                    if s<target:
                        l+=1
                    elif s>target:
                        r-=1
                    else:
                        ans.append([nums[i],nums[j],nums[l],nums[r]])
                        l+=1
                        while l<r and nums[l]==nums[l-1]:
                            l+=1
                        r-=1
                        while l<r and nums[r]==nums[r+1]:
                            r-=1
                j+=1
            i+=1
       
        return ans